src/core/Entry.cpp

```diff
@@ -347,74 +345,35 @@ const CustomData* Entry::customData() const
 
 bool Entry::hasTotp() const
 {
-    return m_attributes->hasKey("TOTP Seed") || m_attributes->hasKey("otp");
+    return !m_data.totpSettings.isNull();
 }
 
 QString Entry::totp() const
 {
     if (hasTotp()) {
-        QString seed = totpSeed();
-        quint64 time = QDateTime::currentDateTime().toTime_t();
-        QString output = Totp::generateTotp(seed.toLatin1(), time, m_data.totpDigits, m_data.totpStep);
-
-        return QString(output);
+        return Totp::generateTotp(m_data.totpSettings);
     }
     return {};
 }
 
-void Entry::setTotp(const QString& seed, quint8& step, quint8& digits)
+void Entry::setTotp(QSharedPointer<Totp::Settings> settings)
 {
     beginUpdate();
-    if (step == 0) {
-        step = Totp::defaultStep;
-    }
-
-    if (digits == 0) {
-        digits = Totp::defaultDigits;
-    }
-    QString data;
-
-    const Totp::Encoder& enc = Totp::encoders.value(digits, Totp::defaultEncoder);
+    m_data.totpSettings = settings;
 
+    auto text = Totp::writeSettings(m_data.totpSettings);
     if (m_attributes->hasKey("otp")) {
-        data = QString("key=%1&step=%2&size=%3").arg(seed).arg(step).arg(enc.digits == 0 ? digits : enc.digits);
-        if (!enc.name.isEmpty()) {
-            data.append("&enocder=").append(enc.name);
-        }
-        m_attributes->set("otp", data, true);
+        m_attributes->set("otp", text, true);
     } else {
-        m_attributes->set("TOTP Seed", seed, true);
-        if (!enc.shortName.isEmpty()) {
-            data = QString("%1;%2").arg(step).arg(enc.shortName);
-        } else {
-            data = QString("%1;%2").arg(step).arg(digits);
-        }
-        m_attributes->set("TOTP Settings", data);
+        m_attributes->set("TOTP Seed", m_data.totpSettings->key, true);
+        m_attributes->set("TOTP Settings", text);
```

Comment:
I wonder if we could make this translatable somehow. But I think you should at least use a static symbol for this instead of repeating the raw string everywhere.

Reply:
I would not translate an attribute key. That would also break compat with the KeePass2 plugins.